### PR TITLE
Node permissioning

### PR DESCRIPTION
**Pull request by ssudhir123**

### CODE REVIEW

File: permissioned-nodes.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  "enode://920b930bcd285cda9edf4c0c92a65026cabfb18d820d121178dacef8902b93203b87b8f5acd01715ff3864cbff041d69affd739e2521d81ac1e40c6b17d25ee@192.168.0.105:30303",
+  "enode://a311f122aa3b2d9deffc721f39fa9ee4816dcec68e7845b5a1c6875ecaf4fcadec48c0b6a6e50bda78e44bd68cf966a3c92509c4abe09fe32b282cae9d1d1d1b@192.168.0.105:30304",
+  "enode://58a71648e004675ae30f6e0953cc89cbbd05a5f6744d2fd15cc04b8923d4079185acefb669d608081b00aa5bfd03722f4843b4c71d9021ff8cb3a25b95673773@192.168.0.105:30305",
+  "enode://aaae04bb274dabf04f03a18548bd763c999ac88d029a756ad69e8082178d6dc10d3ed5be50947ee57478f4c5ae8c73dc88774b2992c942477c589e50a826b0d6@192.168.0.105:30306",
+  "enode://9287b30ed6753eb27a017daa4f23504979070d3353de2a5dc572f46ab338393ee8c5afe00b79d28e8bffab0774c4ecdcfb62e5ac5a51c4544e350fae309fae21@192.168.0.105:30307",
+  "enode://ce70feafc1ca25d815b492b2625625c2a79c7fe551a75ea42a9b28e5c956b117eab0fb86e3f344b51ef8fcdef0d7c3c4305d89089d1b8fbbafd52607679f4475@192.168.0.105:30308",
+  "enode://66f5ad6594aadb6cc7111b033304c3e03a5bf274d2535a3c15613125d35f3a489426569ae587f2d55deffcfda93766fe264fc139f6601ef1932d7113278f27e3@192.168.0.105:30309"
+
+]
```
Comment (chetan) [inline permissioned-nodes.json:13]: I guess this file is meant as a sample? Is there a better place to put it?

File: p2p/permissions.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		if err != nil {
+			glog.V(logger.Error).Infof("Node URL %s: %v\n", url, err)
+			continue
+		}
+		nodes = append(nodes, node)
+	}
+	return nodes
+}
+
+func main() {
```
Comment (chetan) [inline p2p/permissions.go]: This `main()` method looks to be left over from testing.. remove

**Review by chetan — CHANGES_REQUESTED** [review]
Looks good for the most part, see minor comments. Will test tomorrow and update again

File: README.md [hunk 3]
```diff
@@ -335,8 +335,21 @@ QUORUM OPTIONS:
   --singleblockmaker		    Indicate this node is the only node that can create blocks
   --minblocktime value		    Set minimum block time (default: 3)
   --maxblocktime value		    Set max block time (default: 10)
+  --enableNodePermission      Enables node permissioning. If true, it reads from <datadir>/permissioned-nodes.json and allowws only those nodes to connect
```
Comment (patrickmn) [inline README.md]: Rename to --permissioned

File: p2p/permissions.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	// NODE_NAME_LENGTH    = 32
+	PERMISSIONED_CONFIG = "permissioned-nodes.json"
+)
+
+// check if a given node is permissioned to connect to the change
+func isNodePermissioned(nodename string, currentNode string, datadir string, direction string) bool {
+
+	// nodename = nodename[0:NODE_NAME_LENGTH]
+	var permissonedList []string
+	nodes := parsePermissionedNodes(datadir)
```
Comment (patrickmn) [inline p2p/permissions.go:23]: This shouldn't happen every time there's a check. Can e.g. "Permissioner" be a service that's spun up and does the parsing once at startup?

**Review by patrickmn — CHANGES_REQUESTED** [review]

**Comment by patrickmn** [comment]
Generally looks good, I'd say just change the flag name to permissioned, remove main() and change the contents of the sample file. Making it a service/not re-parsing every request can happen later.

File: README.md [hunk 5]
```diff
@@ -335,19 +335,29 @@ QUORUM OPTIONS:
   --singleblockmaker		    Indicate this node is the only node that can create blocks
   --minblocktime value		    Set minimum block time (default: 3)
   --maxblocktime value		    Set max block time (default: 10)
-  --enableNodePermission      Enables node permissioning. If true, it reads from <datadir>/permissioned-nodes.json and allowws only those nodes to connect
+  --permission      Enables node permissioning. If true, it reads from <datadir>/permissioned-nodes.json and allowws only those nodes to connect
```
Comment (patrickmn) [inline README.md:338]: --permissioned, also note the spacing

File: README.md [hunk 6]
```diff
@@ -335,19 +335,29 @@ QUORUM OPTIONS:
...
 ```
 
 Node Permissioning:
 
-Node Permissioning is controlled by a CLI flag --enableNodePermission:
+Node Permissioning is a features that controls which nodes can connect to a given node and also to which nodes this node can dial out to. Currently, it is managed at individual node level by a command line flag "--permission" while starting the node.
```
Comment (patrickmn) [inline README.md:343]: feature
"to which other nodes this one will connect"
at an individual node level
--permissioned

File: README.md [hunk 7]
```diff
@@ -335,19 +335,29 @@ QUORUM OPTIONS:
...
 
-If enabled, the node performs a validation check based on the remote key of the incoming and dialed out.
+If the "--permission" node is present, the node looks for a file named "<data-dir>/permissioned-nodes.json". This file contains the list of enodes specified that this node can connect to and also accept connections only from those nodes. In other words, if permissioning is enabled, only the nodes that are listed in this become part of the network. It is an error to enable permission but not have the permissioned-nodes.json file under <data-dir>. If the permission is specified, but no nodes are present in this file, then this node can neither connect to any node or would accept any incoming connections.
```
Comment (patrickmn) [inline README.md:345]: --permissioned
If the permission is specified -> If permissioning is enabled

File: README.md [hunk 8]
```diff
@@ -335,19 +335,29 @@ QUORUM OPTIONS:
...
+  "enode://remoteky1@ip2:port2",
+  "enode://remoteky1@ip3:port3",
+]
+
+Sample permissioned-nodes.json file:
+[ 
+"enode://6598638ac5b15ee386210156a43f565fa8c48592489d3e66ac774eac759db9eb52866898cf0c5e597a1595d9e60e1a19c84f77df489324e2f3a967207c047470@127.0.0.1:30300",
+]
+
+In the current release, every node has its own copy of the permissioned-nodes.json. It is understood that if different nodes have different list of remote keys. In future releases, the permissioned nodes list will be moved to a smart contract, thereby keeping the list on chain and one global list of nodes that connect to the network.
```
Comment (patrickmn) [inline README.md:359]: It is understood... -> "This copy should be the same for all nodes in the network"?

**Review by patrickmn — CHANGES_REQUESTED** [review]

File: README.md [hunk 9]
```diff
@@ -335,8 +335,31 @@ QUORUM OPTIONS:
   --singleblockmaker		    Indicate this node is the only node that can create blocks
   --minblocktime value		    Set minimum block time (default: 3)
   --maxblocktime value		    Set max block time (default: 10)
+  --permission      Enables node permissioning. If true, it reads from <datadir>/permissioned-nodes.json and allowws only those nodes to connect
```
Comment (tylobban) [inline README.md:338]: should this be 'permission**ed**', per the Name value of the EnableNodePermissionFlag flag?

Minor spelling mistake in the comment allowws -> allows

File: README.md [hunk 10]
```diff
@@ -335,8 +335,31 @@ QUORUM OPTIONS:
...
 ```
 
+Node Permissioning:
+
+Node Permissioning is a features that controls which nodes can connect to a given node and also to which nodes this node can dial out to. Currently, it is managed at individual node level by a command line flag "--permission" while starting the node.
```
Comment (tylobban) [inline README.md:343]: --permission**ed**

**Comment by patrickmn** [comment]
Merged in https://github.com/jpmorganchase/quorum/pull/15

**Pull request closed without merging**